src/vegasflow/vflowplus.py

```diff
@@ -0,0 +1,258 @@
+"""
+    Implementation of vegas+ algorithm:
+
+    adaptive importance sampling + adaptive stratified sampling
+    from https://arxiv.org/abs/2009.05112
+
+    The main interface is the `VegasFlowPlus` class.
+"""
+from itertools import product
+import numpy as np
+import tensorflow as tf
+
+from vegasflow.configflow import (
+    DTYPE,
+    DTYPEINT,
+    fone,
+    fzero,
+    float_me,
+    int_me,
+    BINS_MAX,
+    BETA,
+    MAX_NEVAL_HCUBE,
+)
+from vegasflow.monte_carlo import wrapper, sampler, MonteCarloFlow
+from vegasflow.vflow import VegasFlow, importance_sampling_digest
+from vegasflow.utils import consume_array_into_indices
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+FBINS = float_me(BINS_MAX)
+
+
+@tf.function(
+    input_signature=[
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+        tf.TensorSpec(shape=[], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPEINT),
+        tf.TensorSpec(shape=[None, None], dtype=DTYPE),
+    ]
+)
+def generate_samples_in_hypercubes(rnds, n_strat, n_ev, hypercubes, divisions):
+    """Receives an array of random numbers 0 and 1 and
+    distribute them in each hypercube according to the
+    number of samples in each hypercube specified by n_ev
+
+    Parameters
+    ----------
+        `rnds`: tensor of random number between 0 and 1
+        `n_strat`: tensor with number of stratifications in each dimension
+        `n_ev`: tensor containing number of samples per hypercube
+        `hypercubes`: tensor containing all different hypercube
+        `divisions`: vegas grid
+
+    Returns
+    -------
+        `x` : random numbers collocated in hypercubes
+        `w` : weight of each event
+        `ind`: division index in which each (n_dim) set of random numbers fall
+        `segm` : segmentantion for later computations
+    """
+    # Use the event-per-hypercube information to fix each random event to a hypercub
+    indices = tf.repeat(tf.range(tf.shape(hypercubes, out_type=DTYPEINT)[0]), n_ev)
+    points = float_me(tf.gather(hypercubes, indices))
+    n_evs = float_me(tf.gather(n_ev, indices))
+
+    # Compute in which division of the importance_sampling grid the points fall
+    xn = tf.transpose(points + rnds) * FBINS / float_me(n_strat)
+
+    ind_xn, x, weights = importance_sampling_digest(xn, divisions)
+
+    # Reweight taking into account the number of events per hypercub
+    final_weights = weights / n_evs
+
+    segm = indices
+    return x, ind_xn, final_weights, segm
+
+
+class VegasFlowPlus(VegasFlow):
+    """
+    Implementation of the VEGAS+ algorithm
+    """
+
+    def __init__(self, n_dim, n_events, train=True, adaptive=False, **kwargs):
+        _ = kwargs.setdefault("events_limit", n_events)
+        super().__init__(n_dim, n_events, train, **kwargs)
```

Comment:
Right now, vegas+ cannot break the integration, correct?
Maybe it is worth considering for later developments of vegas+.

Reply:
Yes, and actually we should add a warning for now instead of doing it silently.
It should be indeed a priority.